src/devices/bus/isa/areplay.cpp

```diff
@@ -0,0 +1,614 @@
+		 {
+                  if( ((data & PORT2_IRQ_ENABLE) > 0 )  )
+		  {
+		     if(m_timer_fired == true)
+		     {	
+			if(m_is_heartbeat_present == false) 
+			   force_irq_to(true);
+		     } 
+		  }
+		 } // old irq state was ZERO.  
+		
+		if( (data & PORT2_IRQ_ENABLE) == 0 ) // driver reactivates stuck IRQ by forcing the bit to 0, then 1
+			force_irq_to(false);
+
+		old_irq_state = (data & PORT2_IRQ_ENABLE) ? true : false; // keep position of statement 
+
+		bank_start = m_current_rom_start + SIZE_ROM_WINDOW;
+		bank_end   = bank_start + SIZE_RAM_BANK - 1;
+
+		m_isa->unmap_bank(bank_start, bank_end);
+
+		m_current_RAM_PAGE = data & 0x1f; // Bits 0..4 
+
+		m_isa->install_bank(bank_start, bank_end , 
+				     "banked_ram", &m_board_ram[ SIZE_RAM_BANK * m_current_RAM_PAGE ]
+			           );
+		break; // (end of) case 1 : RAM bank 
+    }
+}
+
+void isa8_areplay_device::device_timer(emu_timer &timer, device_timer_id tid, int param, void *ptr)
+{
+	switch (tid)
+	{
+		default:
+
+if( ioport("DEBUG")->read() != 0 )
+	printf("\nONE SHOT  **** stopped ****!");
+			m_timer_fired = true;
+
+			force_irq_to(false);
+			break; 
+
+	} // switch (timer ID)
+}
+
+
+// Subroutine 
+// => side effect on global variable: 	m_is_heartbeat_present = flag; 
+
+void isa8_areplay_device::force_irq_to(bool flag)
+{	
+	m_is_heartbeat_present = flag; // flag indicates the IRQ is generated by Action Replay
+
+	if(flag == false)
+		one_shot_timer->adjust(attotime::never); 
+	  else
+		one_shot_timer->adjust(attotime::from_msec(ONE_SHOT_TIMER_DELAY_MS)); 
+	raise_processor_interrupt(m_current_irq_selected, flag);
+}
+
+TIMER_DEVICE_CALLBACK_MEMBER(isa8_areplay_device::heartbeat_timer)
+{
+	if(!m_is_heartbeat_present)
+	{	
+		m_timer_fired = false;
+
+		force_irq_to(true); // side effect: m_is_heartbeat_present = true
+	}
+}
+
+void isa8_areplay_device::raise_processor_interrupt(int ref, bool state)
+{
+			switch(ref)
+	      		{
+			case 2:
+				m_isa->irq2_w(state ? ASSERT_LINE : CLEAR_LINE);
+				break;
+	
+			case 3:
+				m_isa->irq3_w(state ? ASSERT_LINE : CLEAR_LINE);
+				break;
+
+			case 4:
+				m_isa->irq4_w(state ? ASSERT_LINE : CLEAR_LINE);
+				break;
+
+			case 5:
+				m_isa->irq5_w(state ? ASSERT_LINE : CLEAR_LINE);
+				break;
+
+			case 6:
+				m_isa->irq6_w(state ? ASSERT_LINE : CLEAR_LINE);
+				break;
+
+			case 7:
+				m_isa->irq7_w(state ? ASSERT_LINE : CLEAR_LINE);
+				break;
+			}
+}
+
+static INPUT_PORTS_START( areplay_dsw )
+	PORT_START("FREEZE_BUTTON")
+	PORT_DIPNAME(0x01, 0x00, "Freeze (push to activate Action Replay)") PORT_IMPULSE(4)
+	PORT_DIPSETTING(0x00, DEF_STR(Off))
+	PORT_DIPSETTING(0x01, DEF_STR(On))
+
+	PORT_START("SLOMO_SWITCH")
+	PORT_DIPNAME(0x01, 0x00, "Slow motion switch (switch OFF at startup)") PORT_TOGGLE
+	PORT_DIPSETTING(0x00, DEF_STR(Off))
+	PORT_DIPSETTING(0x01, DEF_STR(On))
+
+	PORT_START("ROM_ADDRESS")
+	PORT_DIPNAME( 0xff, 0x00, "16K ROM bank (disable BIOS shadowing; exclude in EMM386)")
```

Comment:
Is this really a set of eight jumpers?  If not, don't give it a mask that covers eight bits.  Also, use CONFSETTING/CONFNAME rather than DIPSETTING/DIPNAME for things that are jumpers, not actual DIP switches.

Also, the names are too long - you need to consider how the UI displays this stuff.  You can't include usage information here.